### ISSUE TITLE
Use toad-scheduler job with Promise.all?

### ISSUE THREAD

**Issue by Blutude** [kibertoad/toad-scheduler#159]
I am trying to schedule a job that is to complete actions for 10 devices.
I get this error when I try to use AsyncTask with Promise.all()
`Argument of type '() => Promise<void[][]>' is not assignable to parameter of type '() => Promise<void>'.
  Type 'Promise<void[][]>' is not assignable to type 'Promise<void>'.
    Type 'void[][]' is not assignable to type 'void'.ts(2345)`

but if I cast the Promise.all() as `any`, then it will build fine, and when I try running the code, things seem to run fine.

Is it recommended to split this up into 10 separate AsyncTasks, or can I run AsyncTask with Promise.all()

```ts
const task = new AsyncTask(
    'live location job', 
    () => {
      return Promise.all(
        allUserVehicles.map(async ([userToken, vehicles]) => {
          const userToken = await getUserTokenByEmail(email); // refreshes token if needed
          return Promise.all(vehicles.map((vehicle) => {
            const vehicleClient = new VehicleClient(vehicle.vehicleID);
            vehicleClient.initializeVehicleClient(userToken).then(() => {
              console.log(`[${new Date().toISOString()}] - getting vehicle live data for Vehicle with Vin [${vehicle.vin}] under user [${userToken.email}]`);
              vehicleClient.getAllLiveVehicleData().then(_data => {
                sendMessage({...},
                }).then(
                  () => console.log(`[${new Date().toISOString()}] - sent vehicle live data for Vehicle with Vin [${vehicle.vin}] under user [${userToken.email}]`)
                )
              })
            })
          }));
        })
      );
    },
    (error: Error) => {
      console.error('error while getting tesla vehicle live data on scheduler', error.message);
      throw Error();
      // TODO: test that this throw error stops service
    }
  );
```

**Comment by kibertoad** [comment]
I need to write a test case for this, will report back.

**Comment by Blutude** [comment]
many thanks!
We are planning to go live with this code tonight so any support is greatly appreciated

**Comment by kibertoad** [comment]
@Blutude I've published 2.1.1 which should no longer require casting to `any`.

That said, your code example looks somewhat scary, using `map` function to map to async function is quite a minefield.

I highly recommend watching https://www.youtube.com/watch?v=XV-u_Ow47s0 to understand the risks of mixing callbacks with async functions.

**Comment by Blutude** [comment]
Thanks @kibertoad for these changes, and for sharing this video.

After watching the video, I would not have thought I am abusing promises here, more than what is needed (in the sense that I am not unnecessarily creating promises where I don't need to wait for things). The nature of the flow requires me to wait for 1 step before executing the next one, and so on, and do that for an array of 10  vehicles (or maybe even later once we scale, 300).

The video explains a lot but it does not necessarily provide an alternate solution. It does mention at some point that **“when you await an async function, that creates 3 promises”. Does removing this one async await that I had and replacing it with a promise chain make things at least better here in performance** (see code below)? At least I am reducing the amount of promises if I am understanding this right (am I?).

**I think the best way of improving the performance of this at scale, is to separate the chained promises execution of these vehicles, into smaller vehicle arrays**, that each get run in their own AsyncTask schedule, and start the schedule of each of those AsyncTasks at separate times. That way the loop of the chained promises becomes a smaller loop. Maybe even have 1 AsyncTask per vehicle? so if we have 300 Vehicles, then **is it feasible to have 300 AsyncTasks scheduled at separate times**? Off the top of your head, is it **preferred to chunk them into 30 AsyncTasks of 10 vehicles each**?


```ts
const teslaTask = new AsyncTask(
    'live location tesla job',
    () => Promise.all(allUserVehicles.map(([email, vehicles]) => {
      // refreshes token if needed
      teslaClient.getUserTokenByEmail(email)
      .then(userToken => Promise.all(vehicles.map(vehicle => {
        // then initializes vehicles client using the token
        const vehicleClient = new VehicleClient(vehicle.vehicleID);
        vehicleClient.initializeVehicleClient(userToken)
        .then(() => {
          // then gets vehicle data using the vehicle client
          logger.info(
            `[${new Date().toISOString()}] - getting tesla vehicle live data for Vehicle with Vin [${
              vehicle.vin
            }] under user [${userToken.email}]`,
          );
          vehicleClient.getAllLiveVehicleData()
          .then((data) => {
            // then sends vehicle data to wialon
            WialonClient.getInstance().sendLocation(vehicle, data)
            .then(() =>
              // then logs that it completed
              logger.info(
                `[${new Date().toISOString()}] - sent tesla vehicle live data for Vehicle with Vin [${
                  vehicle.vin
                }] under user [${userToken.email}] to wialon`,
              ),
            );
          });
        });
      })));
    })),
    (error: Error) => {
      logger.error(`error while running scheduled tesla job: ${error.message}`);
      throw Error();
    },
  );
```

**Comment by kibertoad** [comment]
@Blutude It's not just about performance: note that if you throw inside a callback (e. g. inside .map()), error does not automatically propagate to the parent promise, you get an unhandled rejection instead, which will crash your Node instance. So you have to add try-catch at the very least to handle your errors explicitly within every additional Promise level.

If you want to generate predictable load here, I would recommend using a promise pool, e. g. https://www.npmjs.com/package/@supercharge/promise-pool

Then you could use ToadScheduler for just kickstarting the process, and handle the concurrent execution with proper error handling and configurable rate limiting via the promise pool.

**Comment by Blutude** [comment]
To me it feels like using async await with await Promise.all and try catches is cleaner in terms of readability of the code, but as you advised in the documentation of this package, and as it was advised in the video you linked, I removed the use of most async awaits in this flow to reduce the overall amount of promises.

I did edit the block of code to catch all the different promises in the chain of promises. I guess instead I can use a Promise pool like you mentioned, and have each of those functions throw specific errors, so that I can interpret the error and know what to log in the handleError section.
